src/common/libkvs/treeobj.c

```diff
@@ -0,0 +1,496 @@
+ *  the Lawrence Livermore National Laboratory (cf, AUTHORS, DISCLAIMER.LLNS).
+ *  LLNL-CODE-658032 All rights reserved.
+ *
+ *  This file is part of the Flux resource manager framework.
+ *  For details, see https://github.com/flux-framework.
+ *
+ *  This program is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License as published by the Free
+ *  Software Foundation; either version 2 of the license, or (at your option)
+ *  any later version.
+ *
+ *  Flux is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the terms and conditions of the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+ *  See also:  http://www.gnu.org/licenses/
+\*****************************************************************************/
+
+/* See RFC 11 */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <errno.h>
+#include <string.h>
+#include <assert.h>
+
+#include "treeobj.h"
+#include "src/common/libutil/blobref.h"
+#include "src/common/libutil/base64.h"
+
+static const int treeobj_version = 1;
+
+static int treeobj_unpack (json_t *obj, const char **typep, json_t **datap)
+{
+    json_t *data;
+    int version;
+    const char *type;
+    if (!obj || json_unpack (obj, "{s:i s:s s:o !}",
+                                  "ver", &version,
+                                  "type", &type,
+                                  "data", &data) < 0
+             || version != treeobj_version) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (typep)
+        *typep = type;
+    if (datap)
+        *datap = data;
+    return 0;
+}
+
+int treeobj_validate (json_t *obj)
+{
+    json_t *o, *data;
+    const char *type;
+
+    if (treeobj_unpack (obj, &type, &data) < 0)
+        goto inval;
+    if (!strcmp (type, "valref") || !strcmp (type, "dirref")) {
+        int i, len;
+        if (!json_is_array (data))
+            goto inval;
+        len = json_array_size (data);
+        if (len == 0)
+            goto inval;
+        json_array_foreach (data, i, o) {
+            if (blobref_validate (json_string_value (o)) < 0)
+                goto inval;
+        }
+    }
+    else if (!strcmp (type, "dir")) {
+        const char *key;
+        if (!json_is_object (data))
+            goto inval;
+        json_object_foreach (data, key, o) {
+            if (treeobj_validate (o) < 0)
+                goto inval;
+        }
+    }
+    else if (!strcmp (type, "symlink")) {
+        if (!json_is_string (data))
+            goto inval;
+    }
+    else if (!strcmp (type, "val")) {
+    }
```

Comment:
I think that there can be a ```json_is_string()``` check here?  B/c we are base64 encoding, the data should always be a string?